Kernel/VirtIO/VirtIOConsole.h

```diff
@@ -1,53 +1,78 @@
 /*
+ * Copyright (c) 2021, Kyle Pereira <hey@xylepereira.me>
  * Copyright (c) 2021, the SerenityOS developers.
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #pragma once
 
-#include <Kernel/Devices/CharacterDevice.h>
 #include <Kernel/VM/RingBuffer.h>
 #include <Kernel/VirtIO/VirtIO.h>
+#include <Kernel/VirtIO/VirtIOConsolePort.h>
 
 namespace Kernel {
+class VirtIOConsole
+    : public VirtIODevice
+    , public RefCounted<VirtIOConsole> {
+    friend VirtIOConsolePort;
 
-#define VIRTIO_CONSOLE_F_SIZE (1 << 0)
-#define VIRTIO_CONSOLE_F_MULTIPORT (1 << 1)
-#define VIRTIO_CONSOLE_F_EMERG_WRITE (1 << 2)
-
-#define RECEIVEQ 0
-#define TRANSMITQ 1
-
-class VirtIOConsole final : public CharacterDevice
-    , public VirtIODevice {
 public:
     VirtIOConsole(PCI::Address);
-    virtual ~VirtIOConsole() override;
+    virtual ~VirtIOConsole() override = default;
 
-    virtual const char* purpose() const override { return class_name(); }
+    virtual const char* purpose() const override { return "VirtIOConsole"; }
+
+    unsigned device_id() const
+    {
+        return m_device_id;
+    }
 
 private:
-    constexpr static size_t RINGBUFFER_SIZE = 2 * PAGE_SIZE;
-    virtual const char* class_name() const override { return m_class_name.characters(); }
+    enum class ControlEvent : u16 {
+        DeviceReady = 0,
+        DeviceAdd = 1,
+        PortReady = 3,
+        ConsolePort = 4,
+        PortOpen = 6,
+    };
+    struct [[gnu::packed]] ControlMessage {
+        u32 id;
+        u16 event;
+        u16 value;
+
+        enum class Status : u16 {
+            Success = 1,
+            Failure = 0
+        };
 
-    virtual bool can_read(const FileDescription&, size_t) const override;
-    virtual KResultOr<size_t> read(FileDescription&, u64, UserOrKernelBuffer&, size_t) override;
-    virtual bool can_write(const FileDescription&, size_t) const override;
-    virtual KResultOr<size_t> write(FileDescription&, u64, const UserOrKernelBuffer&, size_t) override;
+        enum class PortStatus : u16 {
+            Open = 1,
+            Close = 0
+        };
+    };
 
-    virtual mode_t required_mode() const override { return 0666; }
+    constexpr const static u16 CONTROL_RECEIVEQ = 2;
+    constexpr const static u16 CONTROL_TRANSMITQ = 3;
+    constexpr const static size_t CONTROL_MESSAGE_SIZE = sizeof(ControlMessage);
+    constexpr const static size_t CONTROL_BUFFER_SIZE = CONTROL_MESSAGE_SIZE * 32;
```

Comment:
`constexpr` implies `const`, but if you want to write `const` for some reason, it has to go after the type (i.e. east const).